src/bin/server.rs

```diff
@@ -66,19 +94,43 @@ fn main() {
 
     println!("listening on port {}", port);
 
+    // Give tokio a chance to spawn the first worker thread
+    thread::sleep(Duration::from_millis(10));
+
     // Creating this file tells heroku to tell nginx that the application is ready
     // to receive traffic.
     if heroku {
+        println!("Writing to /tmp/app-initialized");
         File::create("/tmp/app-initialized").unwrap();
     }
 
-    if let Hyper(server) = server {
-        let addr = ([127, 0, 0, 1], port).into();
-        server.run(addr);
-    } else {
-        // Civet server is already running, but we need to block the main thread forever
-        // TODO: handle a graceful shutdown by just waiting for a SIG{INT,TERM}
-        let (_tx, rx) = channel::<()>();
-        rx.recv().unwrap();
+    // Block the main thread until the server has shutdown
+    match server {
+        Hyper(rt) => rt.shutdown_on_idle().wait().unwrap(),
+        Civet(server) => {
+            let (tx, rx) = channel::<()>();
+            ctrlc_handler(move || tx.send(()).unwrap_or(()));
+            rx.recv().unwrap();
+            drop(server);
```

Comment:
The drop seems unnecessary given it'll be dropped at the end of the block which immediately follows.